### ISSUE TITLE
Forest DB crash when device is kept idle for 15 hrs

### ISSUE THREAD

**Issue by msanilkumar2020** [couchbase/couchbase-lite-ios#1554]
Idle testing crash.

Exception Type: EXC_BREAKPOINT (SIGTRAP)
Exception Codes: 0x0000000000000001, 0x000000018128014c
Termination Signal: Trace/BPT trap: 5
Termination Reason: Namespace SIGNAL, Code 0x5
Terminating Process: exc handler [0]
Triggered by Thread: 15
Application Specific Information:
BUG IN CLIENT OF LIBPLATFORM: os_lock is corrupt
Abort Cause 313103321
.
.
Thread 15 name: CouchbaseLite
Thread 15 Crashed:
0 libsystem_platform.dylib 0x000000018128014c _os_lock_corruption_abort + 36
1 libsystem_platform.dylib 0x0000000181280128 _os_lock_corruption_abort + 0
2 xxxxxxxxxx 0x00000001005843f0 update_file_pointers(filemgr*) (filemgr.cc:1431)
3 xxxxxxxxxx 0x0000000100584100 filemgr_close (filemgr.cc:1516)
4 xxxxxxxxxx 0x00000001005b7670 _fdb_close (forestdb.cc:7634)
5 xxxxxxxxxx 0x00000001005b4b04 fdb_check_file_reopen (forestdb.cc:2949)
6 xxxxxxxxxx 0x00000001005b9120 fdb_get (forestdb.cc:3073)
7 xxxxxxxxxx 0x0000000100521564 cbforest::KeyStore::read(cbforest::Document&, cbforest::KeyStore::contentOptions) const (KeyStore.cc:80)
8 xxxxxxxxxx 0x0000000100521468 cbforest::KeyStore::get(cbforest::slice, cbforest::KeyStore::contentOptions) const (KeyStore.cc:61)
9 xxxxxxxxxx 0x000000010051fe0c c4raw_get (c4Database.cc:502)
10 xxxxxxxxxx 0x00000001004fe454 -[CBL_ForestDBStorage getLocalDocumentWithID:revisionID:] (CBL_ForestDBStorage.mm:790)
11 xxxxxxxxxx 0x000000010077d950 -[CBLDatabase existingLocalDocumentWithID:] (CBLDatabase.m:372)
12 xxxxxxxxxx 0x000000010073fcdc -[CBLRestReplicator getLastSequenceFromLocalCheckpointDocument] (CBLRestReplicator.m:673)
13 xxxxxxxxxx 0x000000010073fe30 -[CBLRestReplicator fetchRemoteCheckpointDoc] (CBLRestReplicator.m:690)
14 xxxxxxxxxx 0x000000010073ed54 -[CBLRestReplicator goOnline] (CBLRestReplicator.m:422)
15 xxxxxxxxxx 0x00000001007416d0 -[CBLRestPuller goOnline] (CBLRestPuller.m:191)
16 xxxxxxxxxx 0x000000010073e4c4 -[CBLRestReplicator start] (CBLRestReplicator.m:310)
17 xxxxxxxxxx 0x00000001008140ac -[CBL_Router(Handlers) do_POST_replicate] (CBL_Router+Handlers.m:423)
18 xxxxxxxxxx 0x000000010080e924 -[CBL_Router route] (CBL_Router.m:491)
19 xxxxxxxxxx 0x000000010080ee88 -[CBL_Router run] (CBL_Router.m:520)
20 xxxxxxxxxx 0x000000010081023c __19-[CBL_Router start]_block_invoke (CBL_Router.m:742)
21 Foundation 0x0000000182cd765c __NSThreadPerformPerform + 340
22 CoreFoundation 0x000000018219c278 _CFRUNLOOP_IS_CALLING_OUT_TO_A_SOURCE0_PERFORM_FUNCTION_ + 24
23 CoreFoundation 0x000000018219bbc0 __CFRunLoopDoSources0 + 524
24 CoreFoundation 0x00000001821997c0 __CFRunLoopRun + 804
25 CoreFoundation 0x00000001820c8048 CFRunLoopRunSpecific + 444
26 Foundation 0x0000000182bd9a8c -[NSRunLoop(NSRunLoop) runMode:beforeDate:] + 304
27 xxxxxxxxxx 0x000000010073bba8 -[CBL_RunLoopServer runServerThread] (CBL_Server.m:193)
28 Foundation 0x0000000182cd747c _NSThreadstart_ + 1024
29 libsystem_pthread.dylib 0x0000000181284850 _pthread_body + 240
30 libsystem_pthread.dylib 0x0000000181284760 _pthread_body + 0
31 libsystem_pthread.dylib 0x0000000181281dac thread_start + 4

--------
* Version:   VERSION
* Client OS: NAME_AND_VERSION
* Server:    NAME_AND_VERSION

**Comment by snej** [comment]
This looks like a ForestDB bug; it's not managing its lock correctly when recovering from an error reading from the DB.

The underlying cause is probably that the device is locked and the app's file protection entitlements specify that files can't be read when the device is locked. This will trigger an error from any database access.

**Comment by djpongh** [comment]
Created forestdb issue:
 https://issues.couchbase.com/browse/MB-22538

**Comment by msanilkumar2020** [comment]
@snej / @pasin 
Even in CBL 1.4 we are getting this crash very frequently. When can we get crash fix.

If we forest DB ticket, they already closed this issue.
https://issues.couchbase.com/browse/MB-22538

**Comment by pasin** [comment]
The ForestDB [fix](https://github.com/couchbase/forestdb/commit/a7ccb3ffed27daf65a4b99fb1f942954fabdb524) was merged into the stable branch on 2/1. The fix didn't make it into 1.4.0. We need to update CBForest to pickup the fix.

**Comment by pasin** [comment]
Reopen the issue to pick up the fix.

**Comment by pasin** [comment]
@msanilkumar2020 Thanks for letting us know.

**Comment by msanilkumar2020** [comment]
@pasin , Do you have any update on this issue, this crash is occurring very frequently for us.

**Comment by djpongh** [comment]
@msaniljumar2020 can you try building ForestDB yourself and see if it fixes your issue?

**Comment by msanilkumar2020** [comment]
@djpongh ,

I am already having updated forestDB code, still i am seeing same issue.

I have updated forest DB code by using below command,

git submodule update --init --recursive

Thanks,
Anil Kumar

**Comment by msanilkumar2020** [comment]
Today's crash stack,

Date/Time:           2017-04-03 14:48:25.3230 +0530
Launch Time:         2017-04-03 11:39:02.2266 +0530
OS Version:          iPhone OS 10.3.2 (14F5065b)
Report Version:      104

Exception Type:  EXC_BREAKPOINT (SIGTRAP)
Exception Codes: 0x0000000000000001, 0x000000018f4b50dc
Termination Signal: Trace/BPT trap: 5
Termination Reason: Namespace SIGNAL, Code 0x5
Terminating Process: exc handler [0]
Triggered by Thread:  10

Application Specific Information:
BUG IN CLIENT OF LIBPLATFORM: os_lock is corrupt
Abort Cause 48055239

Filtered syslog:
None found

Thread 10 name:  CouchbaseLite
Thread 10 Crashed:
0   libsystem_platform.dylib      	0x000000018f4b50dc _os_lock_corruption_abort + 36
1   libsystem_platform.dylib      	0x000000018f4b50b8 _os_lock_corruption_abort + 0
2   Push To Talk                  	0x0000000100559644 filemgr_close + 792
3   Push To Talk                  	0x000000010057305c _fdb_close + 128
4   Push To Talk                  	0x0000000100571614 fdb_check_file_reopen + 172
5   Push To Talk                  	0x0000000100573978 fdb_get + 316
6   Push To Talk                  	0x0000000100533d58 cbforest::KeyStore::get(cbforest::slice, cbforest::KeyStore::contentOptions) const (KeyStore.cc:53)
7   Push To Talk                  	0x00000001005334ac c4raw_get (Document.hh:50)
8   Push To Talk                  	0x0000000100525d00 -[CBL_ForestDBStorage getLocalDocumentWithID:revisionID:] (CBL_ForestDBStorage.mm:791)
9   Push To Talk                  	0x00000001007b23c8 -[CBL_Router(Handlers) do_GET:docID:] (CBL_Router+Handlers.m:609)
10  Push To Talk                  	0x00000001007ab718 -[CBL_Router route] (CBL_Router.m:491)
11  Push To Talk                  	0x00000001007abc7c -[CBL_Router run] (CBL_Router.m:520)
12  Push To Talk                  	0x00000001007ad094 __19-[CBL_Router start]_block_invoke (CBL_Router.m:745)
13  Foundation                    	0x0000000190eea4bc __NSThreadPerformPerform + 340
14  CoreFoundation                	0x00000001903a5424 __CFRUNLOOP_IS_CALLING_OUT_TO_A_SOURCE0_PERFORM_FUNCTION__ + 24
15  CoreFoundation                	0x00000001903a4d94 __CFRunLoopDoSources0 + 540
16  CoreFoundation                	0x00000001903a29a0 __CFRunLoopRun + 744
17  CoreFoundation                	0x00000001902d2d94 CFRunLoopRunSpecific + 424
18  Foundation                    	0x0000000190decd64 -[NSRunLoop(NSRunLoop) runMode:beforeDate:] + 304
19  Push To Talk                  	0x00000001006d8580 -[CBL_RunLoopServer runServerThread] (CBL_Server.m:193)
20  Foundation                    	0x0000000190eea2c8 __NSThread__start__ + 996
21  libsystem_pthread.dylib       	0x000000018f4b968c _pthread_body + 240
22  libsystem_pthread.dylib       	0x000000018f4b959c _pthread_body + 0
23  libsystem_pthread.dylib       	0x000000018f4b6cb4 thread_start + 4

**Comment by pasin** [comment]
@msanilkumar2020, only `git submodule update --init --recursive` is not enough. You will need to pull the update from the ForestDB also.

Anyway, we plan to uptake the ForestDB fix soon. I will let you know here when that is done so that you can just update the submodule.

**Comment by ddash1** [comment]
@djpongh @pasin 
Can you please advise if the steps that @msanilkumar2020  mentioned are correct or not?
If correct, does the stack trace prove that the issue is not resolved. Please advise.

Thx

**Comment by pasin** [comment]
@ddash1 @msanilkumar2020 the steps that you updated the forestdb was not completed. I talked to the team, and we will update the forestDB on the current couchbase-lite-ios master branch. I will provide the update here once it is done.

**Comment by msanilkumar2020** [comment]
@pasin ,
Do we have latest forestDb code in branch?

**Comment by pasin** [comment]
Updating the forestdb has a side effect so at this moment we couldn't update.

https://github.com/couchbase/couchbase-lite-ios/issues/1720

**Comment by pasin** [comment]
Can you migrate to SQLite? ForestDB is still in developer preview mode, and we plan to drop forestdb in 2.0. Do you have a reason to continue using forestdb?

**Comment by msanilkumar2020** [comment]
@pasin , as per Couch team suggestion we migrated to ForestDb.

https://blog.couchbase.com/mobile-1-2-enable-forestdb/

Benefits

Here are some of the performance improvements of the new storage system for Couchbase Lite 1.2:

Faster speed — this will vary by app, but we’ve seen from 2x to 5x improvement in tasks like document insertion and querying.
Better concurrency — a writer thread (like the replicator) never blocks readers (like your GUI thread.) This helps keep your application responsive.
Lower RAM footprint — data caches are shared between threads, reducing the overhead of having multiple threads access a database, as occurs during replication.
Auto Database compaction — Database compaction (which frees up disk space) runs in the background without blocking other threads, and is scheduled automatically whenever there’s enough space to recover.

**Comment by snej** [comment]
Apologies if that blog post didn't make it clear enough that ForestDB isn't as supported as SQLite. We never committed to moving exclusively to ForestDB, and we've decided to stop using it in the future (version 2.0) since we've been able to get excellent performance out of SQLite, and SQLite also offers features we need like full-text search and a high-level query engine.

Have you released your app to customers yet? If not, there shouldn't be much impact on you: just remove the line of code that sets the database's storage format to ForestDB.
